### ISSUE TITLE
expose register_reader and register_writer so that manual futures can be implemented

### ISSUE THREAD

**Issue by dvc94ch** [smol-rs/async-io#36]
since rust doesn't support self referential structs there needs to be an alternative to `async fn next(&self)`. Since self isn't owned there is no way to get it into a struct so that it can be polled.

**Comment by stjepang** [comment]
How about this API, would that work for you?

```rust
impl<T> Async<T> {
    fn poll_readable(&mut self, cx: &mut Context<'_>) -> Poll<io::Result<()>>;
    fn poll_writable(&mut self, cx: &mut Context<'_>) -> Poll<io::Result<()>>;
}
```

Note sure about `&mut self` vs `self: Pin<&mut Self>` though...

**Comment by dvc94ch** [comment]
yes, that would work, as it registers a waker. the problem is that at the moment there is no way to get a manual future implementation to wake up.

**Comment by dvc94ch** [comment]
Closed by #38
